### ISSUE TITLE
Matrix multiplication in SystemML on Spark

### ISSUE THREAD

**Issue by Myasuka** [SparkTC/systemml#47]
I run systemml on Spark standalone cluster(not systemml standalone mode), and use simple script below to run a simple matrix-matrix multiplication:
```shell
A = rand(rows=15000, cols=25000, min=0.0, max=1.0, pdf="uniform", sparsity=1.0)
B = rand(rows=25000, cols=15000, min=0.0, max=1.0, pdf="uniform", sparsity=1.0)
C = A %*% B
print( " C[1,1]:" + as.scalar(C[1,1]))
```
I fond systemml would use `mapmm` method to deal with this multiplication ,which seems to broadcast one matrix across the cluster, I notice this performs actually not so well. I have several questions below:

1. Is there any method to assign specific method to deal with matrix-matrix multiplication, such as `cpmm` or `rmm` which is expounded in [Systemml-paper, 2011](http://ieeexplore.ieee.org/xpls/abs_all.jsp?arnumber=5767930).
1. When I use function `rand()` to generate matrix in memory,  to increase the parallelism, is there any method to assign some specific num of partitions?
1. Can systemml call native library such as ATLAS during matrix computation?

**Comment by shirisht** [comment]
Hi Myasuka,

1. CPMM and RMM are only two of the many alternative plans evaluated by SystemML compiler. Whenever one of the input matrix fits entirely in executor's broadcast memory, SystemML chooses mapMM plan to perform matrix multiplication. On the other hand, CPMM and RMM rely on shuffling of input data. Therefore, they are likely to be less efficient than MapMM.

2. NO, such a specification is not allowed because SystemML adopts a declarative approach. As a user, you should never worry about low-level details like partitioning and parallelism. SystemML's compiler determines the best execution strategy (including efficient data distribution) based on data set characteristics and cluster configuration.

3. Out of the box, No. SystemML has its own implementation of (distributed) matrix algebra, which is more efficient than most publicly available libraries.

**Comment by Myasuka** [comment]
Thank for your reply.
I notice there exists many methods to deal with basic matrix-matrix multiplication, can you explain them in brief?

It seems the dimension of sub-blocks during matrix-matrix multiplication is 1000*1000. From your previous statement, it's impossible to change the dimension?

Thanks

**Comment by Myasuka** [comment]
BTW, I set dimension of matrix A and B as `30000x30000` and `30000x30000`, respectively. Systemml still use `MapMM` method to deal with such situation, and the BlockManager in driver could not handle such large matrix as it consumes a lot of memory to store one of the matrix. On the other hand, broadcast such large matrix across the cluster also takes much time.
Would you please kindly explain in which situation that systemml will use other methods such as `PMM`, `CPMM`, `RMM` or `ZIPMM` to deal with matrix-matrix multiplication?

**Comment by mboehm7** [comment]
@Myasuka could you please share the details of your experiment (cluster setup, driver/executor memory budgets, measured runtime, etc)? Note that this particular usecase is compute-intensive since we have 54TFlop of workload, compared to 7.2GB for each of the input matrices. To answer your question, we would fall back to cpmm/rmm if A or B does not fit twice in the driver and once in the remaining executor memory after substracting storage and shuffle memory budget (see spark.storage.memoryFraction, spark.shuffle.memoryFraction) - so although not recommended you can modify the execution plan by changing the memory budgets. An overview of our matrix multiplications can be found here: https://github.com/SparkTC/systemml/blob/master/system-ml/docs/devdocs/MatrixMultiplicationOperators.txt. May I also ask about your use case where you encounter these large-squared matrix multiplications because this is something very uncommon for the workloads we usually deal with?

**Comment by Myasuka** [comment]
Hi, @mboehm7 , I just look into your different implementations of matrix multiplication, there not exists any real workloads.
From [method-choose code](https://github.com/SparkTC/systemml/blob/master/system-ml/src/main/java/com/ibm/bi/dml/hops/AggBinaryOp.java#L1692) here, I have almost figured out how Systemml chooses method to finish matrix multiplication on Spark. By change `spark.storage.memoryFraction` and 'numreducers' in SystemML-config.xml, I can try other methods such `RMM` or `CPMM`. 

BTW, It seems that the `flatMapToPair` action in `RMM` behaves too poor. I have 12 executors with 50GB memory and 16 cores each other. I set  'numreducers' as 24 so that Systemml could choose `RMM`, but it unexpectedly take 40 minutes to finish two `flatMapToPair` stages without any multiplication actually. I think these two stages just need to `flatMap` blocks, and should not take so much time.

**Comment by mboehm7** [comment]
thanks for the clarification @Myasuka - for the purpose of evaluating matrix multiplication operators you can also set FORCED_MMULT_METHOD in AggBinaryOp. I also played around with this scenario and CPMM/RMM indeed perform suboptimal here. For example, on 6 executors w/ 55GB, 24 cores each, I got 1100s for CPMM. I already added a new experimental method called PMapMM (partitioned map matrix multiply) which has the potential of overcoming the bottlenecks of RMM/CPMM for these compute-intensive scenarios: so far we brought it down to ~500s.  

In any case, we are very open for contributions. SystemML is built with certain workloads in mind and there are scenarios where we do not perform well yet. If you have ideas on improving existing matrix multiplication operators or adding new operators, feel free to give it a try and integrate them into SystemML. Let us know if you're interested -  we would point you to the right places where to introduce your changes.

**Comment by Myasuka** [comment]
Thank you for your tips @mboehm7 

I modified the code so that I can configure the matrix-multi method in SystemML-config.xml.

BTW, I have figured out why the "`flatMapToPair`" action takes so much time in `RMM`, becasue `defaultblocksize` is set only `1000` which results in generating `colsB/1000 * |A| + rowsA/1000 * |B|` data during shuffle. By increase `defaultblocksize` can help improve the performance.

**Comment by mboehm7** [comment]
Yes, the additional shuffle is one of the drawbacks of RMM, while CPMM has a bounded parallelism of colsA/blocksize (30 in this case). Note, that the default blocksize of 1000 is a compromise between many operations. For dense matrices, this leads to blocks of 8MB (common L3 cache size), which performs well for chains of operations with narrow dependencies. Conceptually, the blocksize is just yet another  interesting data flow property (like caching, partitioning, and empty block materialization) - down the road, we will integrate these into automatic optimization. Anyway, if you want to contribute, one direction would be a hybrid between RMM and CPMM in order to reduce the degree of replication.
